File: webhook-signature/nodejs/VerifyPayianoWebhookSignature.js
const crypto = require('crypto')

/**
 * Learn how to verify webhook event sent by Payiano API.
 *
 * @author Payiano Team <info@payiano.com>
 * @see https://docs.payiano.com/api/rest/reference/webhooks/verify-signature
 */

/**
 * Check if value is not empty.
 *
 * @param  {mixed} value
 * @return {bool}
 */
const isValueNotEmpty = (value) => {
  if (value && [Object, Array].includes(value.constructor)) {
    const values = Object.values(value)
    return !!values.length && values.some(isValueNotEmpty)
  } else if (typeof value === 'string') {
    return !!value.replace(/\s+/g, '').length
  } else if (value && typeof value === 'boolean') {
    return [true, false].includes(value)
  } else if ([null, undefined].includes(value)) {
    return false
  }

  return true
}

/**
 * Get flatten payload.
 *
 * @param  {mixed} payload
 * @param  {string} parent
 * @return {object}
 */
const getFlattenPayload = (payload, parent = null) =>
  Object.entries(payload).reduce((carry, [key, value]) => {
    key = parent ? `${parent}.${key}` : key

    if (value && [Object, Array].includes(value?.constructor)) {
      Object.assign(carry, {
        ...getFlattenPayload(value, key)
      })
    } else {
      carry[key] = value
    }

    return carry
  }, {})

/**
 * Get cleaned payload.
 *
 * @param  {mixed} flatten
 * @return {object}
 */
const getCleanedPayload = (flatten) =>
  Object.entries(flatten).reduce((carry, [key, value]) => {
    if (isValueNotEmpty(value)) {
      carry[key] =
        typeof value === 'string' ? value.replace(/[\r\n\s]/g, '') : value
    }

    return carry
  }, {})

/**
 * Get sorted payload.
 *
 * @param  {object} cleaned
 * @return {object}
 */
const getSortedPayload = (cleaned) => {
  const sortedKeys = Object.keys(cleaned).sort()

  return sortedKeys.map((key) => ({
    key,
    value: cleaned[key]
  }))
}

/**
 * Get simplified payload.
 *
 * @param  {object} sorted
 * @return {object}
 */
const getSimplifiedPayload = (sorted) =>
  sorted.map(({ key, value }) => `${key}=${value}`)

/**
 * Get signature text.
 *
 * @param  {object} payload
 * @return {object}
 */
const getSignatureText = (payload) => {
  const flatten = getFlattenPayload(payload)

  const cleaned = getCleanedPayload(flatten)

  const sorted = getSortedPayload(cleaned)

  const simplified = getSimplifiedPayload(sorted)

  return simplified.join('&')
}

/**
 * Get computed signature.
 *
 * @param  {object} payload
 * @param  {string} secret
 * @return {string}
 */
const getComputedSignature = (payload, secret) => {
  const signatureText = getSignatureText(payload)

  return crypto.createHmac('sha256', secret).update(signatureText).digest('hex')
}

/**
 * Check if signature is verified.
 *
 * @param  {object} payload
 * @param  {string} receivedSignature
 * @param  {string} secret
 * @return {bool}
 */
const isVerifiedSignature = (payload, receivedSignature, secret) => {
  const computedSignature = getComputedSignature(payload, secret)

  return crypto.timingSafeEqual(
    Buffer.from(computedSignature),
    Buffer.from(receivedSignature)
  )
}

// Find the following example:
const payloadString =
  '{"webhook_event":{"id":"01j3521znn3b6wderr4vbyq18n","type":"company.created","version":"v1","fired_at":"1722572118554"},"webhook_event_attempt":{"id":"01j354j6nkwh3mdvhs6dsmswt8","sent_at":"1722572118554"},"details":{"data":{"company":{"name":"Pyngy URL Shortenr","avatar":null,"is_active":true,"is_approved":false,"employees_count":0,"owners":[{"name":"Amgad Yassen","position":"CEO","percentage":51.5},{"name":"Kamal Allam","position":"CEO","percentage":48.5}],"description":"A leading company providing\\n solutions for converting lengthy\\n URLs into short ones & simplifying online sharing!","social_urls":{"facebook_url":"https://facebook.com/pyngy","linked_in_url":null}}}}}'

const payload = JSON.parse(payloadString)

const receivedSignature =
  '7159d656803a7136be897193dd70a48ca757786d0fe3531f33a48dc17d995725'

const secret = 'OWlPF9plag9KEtYvw3EM+7UDrgXb84xjZPR2TvzJM1I='

console.log({
  getSignatureText: getSignatureText(payload),
  getComputedSignature: getComputedSignature(payload, secret),
  isVerifiedSignature: isVerifiedSignature(payload, receivedSignature, secret)
})

File: webhook-signature/go/VerifyPayianoWebhookSignature.go
/**
 * Learn how to verify webhook event sent by Payiano API.
 *
 * @author Payiano Team <info@payiano.com>
 * @see https://docs.payiano.com/api/rest/reference/webhooks/verify-signature
 */
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

type VerifyPayianoWebhookSignature struct{}

func (v VerifyPayianoWebhookSignature) isVerifiedSignature(payload map[string]interface{}, receivedSignature string, secret string) bool {
	computedSignature := v.getComputedSignature(payload, secret)
	return hmac.Equal([]byte(computedSignature), []byte(receivedSignature))
}

func (v VerifyPayianoWebhookSignature) getComputedSignature(payload map[string]interface{}, secret string) string {
	signatureText := v.getSignatureText(payload)
	h := hmac.New(sha256.New, []byte(secret))
	h.Write([]byte(signatureText))
	return hex.EncodeToString(h.Sum(nil))
}

func (v VerifyPayianoWebhookSignature) getSignatureText(payload map[string]interface{}) string {
	flatten := v.getFlattenPayload(payload, "")
	cleaned := v.getCleanedPayload(flatten)
	sorted := v.getSortedPayload(cleaned)
	simplified := v.getSimplifiedPayload(sorted)
	return strings.Join(simplified, "&")
}

func (v VerifyPayianoWebhookSignature) getFlattenPayload(payload map[string]interface{}, parent string) map[string]interface{} {
	carry := make(map[string]interface{})
	for key, value := range payload {
		newKey := key
		if parent != "" {
			newKey = parent + "." + key
		}

		switch value := value.(type) {
		case map[string]interface{}:
			flattened := v.getFlattenPayload(value, newKey)
			for k, v := range flattened {
				carry[k] = v
			}
		case []interface{}:
			// Sort the list elements if they contain maps
			for index, item := range value {
				indexedKey := fmt.Sprintf("%s.%d", newKey, index)
				switch item := item.(type) {
				case map[string]interface{}:
					flattened := v.getFlattenPayload(item, indexedKey)
					for k, v := range flattened {
						carry[k] = v
					}
				case []interface{}:
					flattened := v.getFlattenPayload(map[string]interface{}{fmt.Sprintf("%d", index): item}, indexedKey)
					for k, v := range flattened {
						carry[k] = v
					}
				default:
					carry[indexedKey] = item
				}
			}
		default:
			carry[newKey] = value
		}
	}
	return carry
}

func (v VerifyPayianoWebhookSignature) getCleanedPayload(flatten map[string]interface{}) map[string]interface{} {
	carry := make(map[string]interface{})
	for key, value := range flatten {
		if v.isValueNotEmpty(value) {
			switch value := value.(type) {
			case string:
				value = strings.TrimSpace(value)
				value = strings.ReplaceAll(value, "\n", "")
				value = strings.ReplaceAll(value, "\t", "")
				value = strings.ReplaceAll(value, " ", "")
				carry[key] = value
			case bool:
				if value {
					carry[key] = "true"
				} else {
					carry[key] = "false"
				}
			default:
				carry[key] = value
			}
		}
	}
	return carry
}

func (v VerifyPayianoWebhookSignature) getSimplifiedPayload(sorted map[string]interface{}) []string {
	keys := make([]string, 0, len(sorted))
	for key := range sorted {
		keys = append(keys, key)
	}

	// Sort the keys to ensure a consistent order
	sort.Strings(keys)

	carry := make([]string, 0, len(sorted))
	for _, key := range keys {
		carry = append(carry, fmt.Sprintf("%s=%v", key, sorted[key]))
	}

	return carry
}

func (v VerifyPayianoWebhookSignature) getSortedPayload(cleaned map[string]interface{}) map[string]interface{} {
	sorted := make(map[string]interface{})
	keys := make([]string, 0, len(cleaned))
	for key := range cleaned {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		sorted[key] = cleaned[key]
	}
	return sorted
}

func (v VerifyPayianoWebhookSignature) isValueNotEmpty(value interface{}) bool {
	switch value := value.(type) {
	case map[string]interface{}:
		for _, v := range value {
			if v != nil {
				return true
			}
		}
	case string:
		return len(strings.TrimSpace(value)) > 0
	case bool:
		return true
	default:
		return value != nil
	}
	return false
}

func main() {
	payloadString := `{"webhook_event":{"id":"01j3521znn3b6wderr4vbyq18n","type":"company.created","version":"v1","fired_at":"1722572118554"},"webhook_event_attempt":{"id":"01j354j6nkwh3mdvhs6dsmswt8","sent_at":"1722572118554"},"details":{"data":{"company":{"name":"Pyngy URL Shortenr","avatar":null,"is_active":true,"is_approved":false,"employees_count":0,"owners":[{"name":"Amgad Yassen","position":"CEO","percentage":51.5},{"name":"Kamal Allam","position":"CEO","percentage":48.5}],"description":"A leading company providing\n solutions for converting lengthy\n URLs into short ones & simplifying online sharing!","social_urls":{"facebook_url":"https://facebook.com/pyngy","linked_in_url":null}}}}}`

	var payload map[string]interface{}
	json.Unmarshal([]byte(payloadString), &payload)

	receivedSignature := "7159d656803a7136be897193dd70a48ca757786d0fe3531f33a48dc17d995725"
	secret := "OWlPF9plag9KEtYvw3EM+7UDrgXb84xjZPR2TvzJM1I="

	verifier := VerifyPayianoWebhookSignature{}
	fmt.Println(map[string]interface{}{
		"getSignatureText":     verifier.getSignatureText(payload),
		"getComputedSignature": verifier.getComputedSignature(payload, secret),
		"isVerifiedSignature":  verifier.isVerifiedSignature(payload, receivedSignature, secret),
	})
}
